Apps/AccountDataAccess/src/Patient/PatientRepository.cs

```diff
@@ -80,55 +90,113 @@ public async Task<PatientQueryResult> Query(PatientQuery query, CancellationToke
             }
 
-            return apiResult;
+            PatientModel? patient = await this.clientRegistriesDelegate.GetDemographicsAsync(type, identifier, disableIdValidation).ConfigureAwait(true);
+
+            // Only cache if validation is enabled (as some clients could get invalid data) and when successful.
+            if (!disableIdValidation && patient != null)
+            {
+                this.CachePatient(patient);
+            }
+
+            return patient;
         }
 
         private async Task<PatientQueryResult> Handle(PatientDetailsQuery query, CancellationToken ct)
         {
+            this.logger.LogDebug("Patient details query source: {Source}", query.Source);
+
             if (ct.IsCancellationRequested)
             {
                 throw new InvalidOperationException("cancellation was requested");
             }
 
             if (query.Hdid != null)
             {
-                ApiResult<PatientModel> result = await this.GetDemographicsAsync(OidType.Hdid, query.Hdid, PatientIdentifierType.Hdid).ConfigureAwait(true);
-                return new PatientQueryResult(new[] { result.ResourcePayload! });
+                PatientModel? patient = ShouldCheckCache(query.Source) ? this.GetFromCache(query.Hdid, PatientIdentifierType.Hdid) : null;
+
+                if (patient == null && ShouldCheckEmpi(query.Source))
+                {
+                    try
+                    {
+                        patient = await this.GetDemographicsAsync(OidType.Hdid, query.Hdid).ConfigureAwait(true);
+                    }
+                    catch (CommunicationException e)
+                    {
+                        if (ShouldCheckPhsa(query.Source))
+                        {
+                            this.logger.LogError("Will call PHSA for patient due to EMPI Communication Exception when trying to retrieve patient information: {Exception}", e);
+                            patient = await this.GetPatientIdentityAsync(query.Hdid).ConfigureAwait(true);
+                        }
+                        else
+                        {
+                            throw;
+                        }
+                    }
+                }
+
+                return new PatientQueryResult(new[] { patient! });
             }
 
             if (query.Phn != null)
             {
-                ApiResult<PatientModel> result = await this.GetDemographicsAsync(OidType.Phn, query.Phn, PatientIdentifierType.Phn).ConfigureAwait(true);
-                return new PatientQueryResult(new[] { result.ResourcePayload! });
+                PatientModel? patient = (ShouldCheckCache(query.Source) ? this.GetFromCache(query.Phn, PatientIdentifierType.Phn) : null) ??
+                                        await this.GetDemographicsAsync(OidType.Phn, query.Phn).ConfigureAwait(true);
+
+                return new PatientQueryResult(new[] { patient! });
             }
```

Comment:
I guess we don't check PHSA in this case because they don't have an endpoint that takes a PHN.  May be worth confirming this is desired behaviour.